### PR TITLE
Add MCLMC sampler

### PR DESCRIPTION
**Pull request by reubenharry**
Todos: 

- [x] add tuning: without the tuning algorithm, MCLMC is next to useless
 - [x] *If I add a new sampler*, there is an issue discussing it already;
 - [x] We should be able to understand what the PR does from its title only;
 - [x] There is a high-level description of the changes;
 - [x] There are links to *all* the relevant issues, discussions and PRs;
 - [x] The branch is rebased on the latest `main` commit;
 - [x] Commit messages follow these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html);
 - [x] The code respects the current naming conventions;
 - [x] Docstrings follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html)
 - [x] `pre-commit` is installed and configured on your machine, and you ran it before opening the PR;
 - [x] There are tests covering the changes;
 - [x] The doc is up-to-date; (add references to bibliography, check what you need to do for API docs)
 - [x] *If I add a new sampler** I added/updated related [examples](https://github.com/blackjax-devs/blackjax/tree/main/examples)
 - [x]  Remove the `explore.py` script
 - [x] Test that the dE and logdensity reported are correct

As per #530, this PR adds MCLMC (Microcanonical Langevin Monte Carlo) as a sampling method to BlackJax. It is an adaptation of the original repo: https://github.com/JakobRobnik/MicroCanonicalHMC/tree/master.

In particular, this is an implementation of the single chain partially momentum updating algorithm (as opposed to full updates every n steps). It also will implement (once this PR is ready) the MCLMC tuning algorithm.

### CODE REVIEW

File: blackjax/mcmc/integrators.py [hunk 1]
```diff
@@ -246,3 +246,25 @@ def one_step(state: IntegratorState, step_size: float) -> IntegratorState:
...
+        # V T V T V
+        sigma = jax.numpy.sqrt(inverse_mass_matrix)
+        uu, r1 = V(step_size * lambda_c, state.momentum, state.logdensity_grad * sigma)
+        xx, ll, gg = T(step_size, state.position, 0.5 * uu * sigma)
+        uu, r2 = V(step_size * (1 - 2 * lambda_c), uu, gg * sigma)
+        xx, ll, gg = T(step_size, xx, 0.5 * uu * sigma)
+        uu, r3 = V(step_size * lambda_c, uu, gg * sigma)
+
+        # kinetic energy change
+        dim = xx.shape[0]
```
Comment (junpenglao) [inline blackjax/mcmc/integrators.py:265]: Infer the shape from `inverse_mass_matrix` outside the function `step`

File: blackjax/mcmc/integrators.py
Comment (junpenglao) [inline blackjax/mcmc/integrators.py:1]: Splitting the changes here into https://github.com/blackjax-devs/blackjax/pull/589

File: blackjax/adaptation/step_size.py [hunk 2]
```diff
@@ -257,3 +260,279 @@ def update(rss_state: ReasonableStepSizeState) -> ReasonableStepSizeState:
...
+
+class MCLMCAdaptationState(NamedTuple):
+    """Tunable parameters for MCLMC"""
+
+    L: float
+    step_size: float
+
+
+
+def ess_corr(x):
```
Comment (junpenglao) [inline blackjax/adaptation/step_size.py:273]: Instead of copying large proportion of code, you should refactor the origin function.

File: blackjax/mcmc/mclmc.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+import jax
+import jax.numpy as jnp
+
+import blackjax.mcmc.integrators as integrators
+from blackjax.base import SamplingAlgorithm
+from blackjax.types import Array, ArrayLike, PRNGKey
+
+__all__ = ["MCLMCState", "MCLMCInfo", "init", "build_kernel", "mclmc"]
+
+MCLMCState = integrators.IntegratorState
```
Comment (junpenglao) [inline blackjax/mcmc/mclmc.py:26]: Use IntegratorState directly.

**Review by junpenglao — COMMENTED** [review]
Some overall pointer/comments:
- Please use more descriptive naming for variables and function name
- Be careful around the assumption of function input (some of the helper function is not pytree friendly).

File: blackjax/adaptation/step_size.py (hunk 2 above)
Comment (reubenharry) [reply blackjax/adaptation/step_size.py:273]: Yep, will definitely do that.

File: blackjax/mcmc/integrators.py [hunk 4]
```diff
@@ -1,3 +1,4 @@
+# @title `integrators.py` from https://github.com/blackjax-devs/blackjax/pull/589
```
Comment (junpenglao) [inline blackjax/mcmc/integrators.py:1]: revert

Comment (reubenharry) [reply blackjax/mcmc/integrators.py:1]: Can you explain?

Comment (junpenglao) [reply blackjax/mcmc/integrators.py:1]: I mean remove this change (since it was added because you were copying the changes from the integrator PR)

**Comment by reubenharry** [comment]
After a bit of a mammoth effort, the PR is ready for review. 

One possible point of objection is that there are still some poorly named variables, now confined to certain functions in the adaptation code. These I have not renamed because I don't fully understand what all of them are doing, so some reading/talking to Jakob is in order. 

But there's now a test of both the tuning and the kernel, that uses pytrees, showing that the code is pytree generic. Moreover, the implementation reproduces the original MCLMC implementation perfectly (with appropriate choices of random seeds), so other than a few details (I removed the preconditioning for now, for simplicity), it's hopefully looking good.

File: blackjax/adaptation/mclmc_adaptation.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        L (float): The momentum decoherent rate for the MCLMC algorithm.
+        step_size (float): The step size used for the MCLMC algorithm.
+    """
+
+    L: float
+    step_size: float
+
+
+def mclmc_find_L_and_step_size(
+    kernel,
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:41]:
```suggestion
    mclmc_kernel,
```

File: blackjax/adaptation/mclmc_adaptation.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    num_steps,
+    state,
+    part1_key,
+    part2_key,
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:45]: Better to use a single key, and split below.

File: blackjax/adaptation/mclmc_adaptation.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+    return state, params
+
+
+def make_L_step_size_adaptation(
+    kernel,
+    dim,
+    frac_tune1,
+    frac_tune2,
+    varEwanted=1e-3,
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:92]: `target_energy_var`?

File: blackjax/adaptation/mclmc_adaptation.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+    sigma_xi=1.5,
+    num_effective_samples=150,
+):
+    """Adapts the stepsize and L of the MCLMC kernel. Designed for the unadjusted MCLMC"""
+
+    gamma_forget = (num_effective_samples - 1.0) / (num_effective_samples + 1.0)
+
+    def predictor(state_old, params, adaptive_state, rng_key):
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:100]: state_old -> previous_state
state_new -> next_state

Same for below and other files.

File: blackjax/adaptation/mclmc_adaptation.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+        )
+
+        return (state, params_final, adaptive_state, kalman_state), None
+
+    def L_step_size_adaptation(state, params, num_steps, rng_key):
+        num_steps1, num_steps2 = int(num_steps * frac_tune1), int(
+            num_steps * frac_tune2
+        )
+        # L_step_size_adaptation_keys = jax.random.split(rng_key, num_steps1 + num_steps2)
+        L_step_size_adaptation_keys = jnp.array([rng_key] * (num_steps1 + num_steps2))
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:171]: Why using the same key? I see you commented out the previous line

File: blackjax/adaptation/mclmc_adaptation.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+            W + w + zero_prevention
+        )  # Update <f(x)> with a Kalman filter
+        W += w
+        return (W, F1, F2)
+
+    adap0 = (0.0, 0.0, jnp.inf)
+
+    def step(iteration_state, weight_and_key):
+        outer_weight, rng_key = weight_and_key
+        """does one step of the dynamics and updates the estimate of the posterior size and optimal stepsize"""
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:154]: This comment should be 1 line above

File: blackjax/adaptation/mclmc_adaptation.py [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+                kernel(rng_key=k, state=s, L=params.L, step_size=params.step_size)
+            ),
+            init=state,
+            xs=adaptation_L_keys,
+        )
+        samples = info.transformed_position  # tranform is the identity here
+        flat_samples, _ = ravel_pytree(samples)
+        dim = pytree_size(state.position)
+        flat_samples = flat_samples.reshape(-1, dim)
+        ESS = 0.5 * effective_sample_size(
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:220]: Ah because ESS in Stan is designed to work on 1+ chains. In this case, dont repeat the sample, but reshape it into 2 chains (basically split ESS, which is how single chain ESS usually computed)

File: blackjax/adaptation/mclmc_adaptation.py [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+    return L_step_size_adaptation
+
+
+def make_adaptation_L(kernel, frac, Lfactor):
+    """determine L by the autocorrelations (around 10 effective samples are needed for this to be accurate)"""
+
+    def adaptation_L(state, params, num_steps, key):
+        num_steps = int(num_steps * frac)
+        # adaptation_L_keys = jax.random.split(key, num_steps)
+        adaptation_L_keys = jnp.array([key] * (num_steps))
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:206]: Same comment here that i assume different key should be used

File: blackjax/adaptation/mclmc_adaptation.py [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+            init=state,
+            xs=adaptation_L_keys,
+        )
+        samples = info.transformed_position  # tranform is the identity here
+        flat_samples, _ = ravel_pytree(samples)
+        dim = pytree_size(state.position)
+        flat_samples = flat_samples.reshape(-1, dim)
+        ESS = 0.5 * effective_sample_size(
+            jnp.array([flat_samples, flat_samples])
+        )  # TODO: should only use a single chain here
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:222]:
```suggestion
        _, unravel_fn = ravel_pytree(state.position)
        flat_samples = jax.vmap(unravel_fn)(samples)
        flat_samples = flat_samples.reshape(2, num_steps // 2, -1)
        ESS = effective_sample_size(flat_samples)
```

File: blackjax/adaptation/mclmc_adaptation.py [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+        return state, params._replace(
+            L=Lfactor * params.step_size * jnp.average(num_steps / ESS)
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:225]:
```suggestion
            L=Lfactor * params.step_size * jnp.mean(num_steps / ESS)
```

File: blackjax/mcmc/mclmc.py [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+    logdensity :
+        The log-density of the distribution at the current step of the MCLMC chain.
+    dE :
+        The difference in energy between the current and previous step.
+    """
+
+    transformed_position: Array
+    logdensity: float
+    kinetic_change: float
+    dE: float
```
Comment (junpenglao) [inline blackjax/mcmc/mclmc.py:44]:
```suggestion
    energy_diff: float
```

File: blackjax/util.py [hunk 16]
```diff
@@ -104,6 +104,31 @@ def generate_unit_vector(
     return unravel_fn(sample / jnp.linalg.norm(sample))
 
 
+def partially_refresh_momentum(momentum, rng_key, step_size, L):
```
Comment (junpenglao) [inline blackjax/util.py:107]: Move this back to blackjax/mcmc/mclmc.py

File: blackjax/mcmc/mclmc.py [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+    def kernel(
+        rng_key: PRNGKey, state: IntegratorState, L: float, step_size: float
+    ) -> tuple[IntegratorState, MCLMCInfo]:
+        (position, momentum, logdensity, logdensitygrad), kinetic_change = step(
+            state, step_size
+        )
+
+        # dim = position.shape[0]
+        dim = 2
```
Comment (junpenglao) [inline blackjax/mcmc/mclmc.py:89]: dim = 2 is intended?

**Review by junpenglao — COMMENTED** [review]
Great job! I indeed have some comments re naming.

File: blackjax/adaptation/mclmc_adaptation.py (hunk 5 above)
Comment (reubenharry) [reply blackjax/adaptation/mclmc_adaptation.py]: resolved

File: blackjax/adaptation/mclmc_adaptation.py (hunk 6 above)
Comment (reubenharry) [reply blackjax/adaptation/mclmc_adaptation.py]: Resolved

File: blackjax/__init__.py [hunk 18]
```diff
@@ -12,7 +13,9 @@
 from .mcmc.hmc import dynamic_hmc, hmc
 from .mcmc.mala import mala
 from .mcmc.marginal_latent_gaussian import mgrad_gaussian
+from .mcmc.mclmc import mclmc
```
Comment (junpenglao) [inline blackjax/__init__.py:16]: remove (there is 2 same import here.

File: blackjax/__init__.py [hunk 19]
```diff
@@ -4,6 +4,7 @@
 from .adaptation.meads_adaptation import meads_adaptation
 from .adaptation.pathfinder_adaptation import pathfinder_adaptation
 from .adaptation.window_adaptation import window_adaptation
+from .adaptation.mclmc_adaptation import mclmc_find_L_and_step_size
```
Comment (junpenglao) [inline blackjax/__init__.py:7]: If you want to expose this you need to add it in the `__all__` below as well.

File: blackjax/adaptation/mclmc_adaptation.py (hunk 7 above)
Comment (reubenharry) [reply blackjax/adaptation/mclmc_adaptation.py:92]: I think `var` might be for "variance". Currently reading up on the algorithm

File: blackjax/mcmc/mclmc.py (hunk 17 above)
Comment (reubenharry) [reply blackjax/mcmc/mclmc.py]: Definitely not, good catch.

File: blackjax/adaptation/mclmc_adaptation.py (hunk 13 above)
Comment (junpenglao) [reply blackjax/adaptation/mclmc_adaptation.py:222]: Ah right... try:
```python
                flat_samples = jax.vmap(lambda x: ravel_pytree(x)[0])(samples)
	        flat_samples = flat_samples.reshape(2, num_steps // 2, -1)
	        ESS = effective_sample_size(flat_samples)
```

Comment (reubenharry) [reply blackjax/adaptation/mclmc_adaptation.py:222]: This passes the test, but gives a numerically different result (slightly) such that the original MCLMC and blackjax implementation no longer align. Is there a reason effective_sample_size forbids a single chain?

Comment (junpenglao) [reply blackjax/adaptation/mclmc_adaptation.py:222]: Because `effective_sample_size` by design use the within chain and cross chain variance to compute ess, so most implementation would split the single chain into 2 (see https://mc-stan.org/docs/2_18/reference-manual/effective-sample-size-section.html). 
I think it is fine if we dont align here, because this is a more accurate implementation IMO.

Comment (reubenharry) [reply blackjax/adaptation/mclmc_adaptation.py]: OK, updated.

File: blackjax/adaptation/mclmc_adaptation.py [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+    """
+    Finds the optimal value of the parameters for the MCLMC algorithm.
+
+    Args:
+        kernel: The kernel function used for the MCMC algorithm.
+        num_steps: The number of MCMC steps that will subsequently be run, after tuning
+        state: The initial state of the MCMC algorithm.
+        frac_tune1: The fraction of tuning for the first step of the adaptation.
+        frac_tune2: The fraction of tuning for the second step of the adaptation.
+        frac_tune3: The fraction of tuning for the third step of the adaptation.
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:61]: Add docstring for the new kwargs

File: blackjax/adaptation/mclmc_adaptation.py [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+
+    Returns:
+        state: The final state of the MCMC algorithm.
+        params: The final hyperparameters of the MCMC algorithm.
+    """
+    dim = pytree_size(state.position)
+    params = MCLMCAdaptationState(jnp.sqrt(dim), jnp.sqrt(dim) * 0.25)
+    varEwanted = 5e-4
```
Comment (junpenglao) [inline blackjax/adaptation/mclmc_adaptation.py:69]: remove this line (as you are providing as a kwarg now)

File: blackjax/adaptation/mclmc_adaptation.py (hunk 8 above)
Comment (junpenglao) [reply blackjax/adaptation/mclmc_adaptation.py:103]: Reminder to change the naming.

File: blackjax/adaptation/mclmc_adaptation.py (hunk 10 above)
Comment (junpenglao) [reply blackjax/adaptation/mclmc_adaptation.py:156]: Reminder to move the `"""` comment 1 line up.

File: blackjax/mcmc/mclmc.py [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+
+    Returns
+    -------
+    momentum with random change in angle
+    """
+    m, unravel_fn = ravel_pytree(momentum)
+    dim = m.shape[0]
+    nu = jnp.sqrt((jnp.exp(2 * step_size / L) - 1.0) / dim)
+    z = nu * normal(rng_key, shape=m.shape, dtype=m.dtype)
+    return unravel_fn((m + z) / jnp.sqrt(jnp.sum(jnp.square(m + z)))), dim
```
Comment (junpenglao) [inline blackjax/mcmc/mclmc.py:205]:
```suggestion
    return unravel_fn((m + z) / jnp.linalg.norm(m + z)), dim
```

**Review by junpenglao — APPROVED** [review]
Great work!! Thank you
